File: leetcode_python/validBST.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root:
            return True
        return self.dfs(root,-sys.maxint-1, sys.maxint)
        
    def dfs(self,root,min_v,max_v):
        if not root:
            return True
        if root.val >=max_v or root.val <= min_v:
            return False
        return self.dfs(root.left,min_v,root.val) and self.dfs(root.right,root.val,max_v)

File: leetcode_python/validBST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root) {
    let max = Number.MAX_SAFE_INTEGER;
    let min = Number.MIN_SAFE_INTEGER;
    return helper(root,min,max);
    
    function helper(root,min,max){
        if(!root) return true;
        if(root.val <= min) return false;
        if(root.val >= max) return false;
        return helper(root.left,min,root.val) && helper(root.right,root.val,max);
    }
};